File: backend/src/StockFlow.Core.Domain/Entities/Product.cs
using StockFlow.Core.Domain.Common;

namespace StockFlow.Core.Domain.Entities;

public sealed class Product : Entity
{
    public Guid CategoryId { get; private set; }
    public string Name { get; private set; }
    public string Sku { get; private set; }
    public decimal UnitPrice { get; private set; }
    public bool IsActive { get; private set; }

    public Product(Guid categoryId, string name, string sku, decimal unitPrice)
    {
        if (categoryId == Guid.Empty)
        {
            throw new ArgumentException("Category is required.", nameof(categoryId));
        }

        if (unitPrice < 0)
        {
            throw new ArgumentOutOfRangeException(nameof(unitPrice), "Unit price cannot be negative.");
        }

        CategoryId = categoryId;
        Name = string.IsNullOrWhiteSpace(name)
            ? throw new ArgumentException("Product name is required.", nameof(name))
            : name.Trim();
        Sku = string.IsNullOrWhiteSpace(sku)
            ? throw new ArgumentException("SKU is required.", nameof(sku))
            : sku.Trim().ToUpperInvariant();
        UnitPrice = unitPrice;
        IsActive = true;
    }
}

File: mobile/stockflow_mobile/lib/core/models/product.dart
class Product {
  final String id;
  final String categoryId;
  final String name;
  final String sku;
  final double unitPrice;
  final bool isActive;

  Product({
    required this.id,
    required this.categoryId,
    required this.name,
    required this.sku,
    required this.unitPrice,
    required this.isActive,
  });

  factory Product.fromJson(Map<String, dynamic> json) => Product(
        id: json['id'] as String,
        categoryId: json['categoryId'] as String,
        name: json['name'] as String,
        sku: json['sku'] as String,
        unitPrice: (json['unitPrice'] as num).toDouble(),
        isActive: json['isActive'] as bool,
      );
}
